### PR TITLE
Add stable sort implementation to `Slice` and `Array`

### LINKED ISSUE #6057: [Proposal] Array stable sort
It may be useful to have the option to stable sort arrays.

I'm thinking about adding it as a flag in `Array#sort`, defaulting to false, signature looks like this:
```
def sort(stable = false) : Array(T)
```

Therefore if needed, it can provide stable sorting by calling:
`[1,...,-1].sort(stable: true)`

Another option is of course to have a separate method (like in c++ or golang):
`[1,...,-1].stable_sort`

Tbh I find the version with parameter a little bit better - if doesn't add a new method to array public interface and doesn't break anything (as it defaults to false).

In spirit if this comment: https://github.com/crystal-lang/crystal/issues/2350#issuecomment-261742027

I would be very happy to implement either of this versions.

### PR DESCRIPTION
**Pull request by MakeNowJust**
Fixed #6057 

The stable sort algorithm is ported from Rust (it is called `merge_sort` in Rust, even though it is extended.)
It is similar to TimSort, but galloping is dropped.
Of course, it holds good properties of TimeSort.
For example, sorting is done in linear time when the collection is already sorted.

This PR adds a `stable` option to sorting methods (`#sort`, `#sort!`, `#sort_by` and `#sort_by!`) of `Array` and `Slice`, respectively.
The option's default value is `false` for backward compatibility.

```crystal
p (0..20).sort_by(stable: false) { |i| i % 2 }
# => [0, 18, 16, 14, 12, 10, 8, 6, 4, 2, 20, 7, 9, 5, 11, 13, 3, 15, 17, 1, 19]

p (0..20).sort_by(stable: true) { |i| i % 2 }
# => [0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 1, 3, 5, 7, 9, 11, 13, 15, 17, 19]
```

It is my last work in 2020. Have a happy new year 🎉

### CODE REVIEW

File: src/slice/sort.cr
Comment (j8r) [inline src/slice/sort.cr:279]: This is particularly true here: which `size` is used, the method or the variable?

File: src/slice/sort.cr
Comment (j8r) [inline src/slice/sort.cr:404]: `v` is not very descriptive, `other` will be better.

`size` is already a method of `Slice`, a better name would be `other_size`.

File: src/slice/sort.cr
Comment (j8r) [inline src/slice/sort.cr:279]: This is particularly true here: which `size` is used, the method or the variable?

**Review by j8r — COMMENTED** [review]
Great! I'd rename `v` to `other` and `size` to `other_size`.

Having a variable for this is not even necessary, `other.size` will fit as well.

File: src/slice/sort.cr
Comment (kostya) [inline src/slice/sort.cr:559]: this line so hard to read, mix of `if` and ternary operator

File: src/slice/sort.cr
Comment (MakeNowJust) [reply src/slice/sort.cr:404]: The name `v` comes from Rust. I guess it means "**v**ector" which is slice of Crystal in Rust.

And, this variable refers to the slice to be sorted, so I think `other` does not make sense. We can rename it to `slice` for Crystal.

File: src/slice/sort.cr
Comment (MakeNowJust) [reply src/slice/sort.cr:279]: Here is in class method. It is clearly variable.

Comment (j8r) [reply src/slice/sort.cr:279]: right, it is in a class method, and there is no `Slice.size`. So, ok

File: src/slice/sort.cr
Comment (asterite) [reply src/slice/sort.cr:404]: Please let's use the names as in Rust, it'll be easier to apply patches.

Also, feel free to ignore comments unless they are from a core team member.

File: src/slice.cr
Comment (j8r) [inline src/slice.cr:694]:
```suggestion
  def sort!(*, stable : Bool = false, &block : T, T -> U) : Slice(T) forall U
```

File: src/slice.cr
Comment (j8r) [inline src/slice.cr:717]:
```suggestion
  def sort_by(*, stable : Bool = false, &block : T ->) : Slice(T)
```

File: src/slice.cr
Comment (j8r) [inline src/slice.cr:730]:
```suggestion
  def sort_by!(*, stable : Bool = false, &block : T ->) : Slice(T)
```

**Comment by j8r** [comment]
I think the protected methods should be tested too, ideally.

File: src/slice/sort.cr
Comment (straight-shoota) [reply src/slice/sort.cr:404]: > Also, feel free to ignore comments unless they are from a core team member.

That sounds a bit harsh, doesn't it? Reviews and suggestions from everyone are welcome, but it's usually advisable to wait for other comments, especially from core team members (because that's the ones who have to approve a PR).

Comment (asterite) [reply src/slice/sort.cr:404]: Truth sometimes is harsh, yes.

Comment (asterite) [reply src/slice/sort.cr:404]: Well, the truth is that some comments don't need to be addressed yet people don't know that and they lose time reverting or discussing those things. And some are simply plain wrong and I have to lose time jumping to address them to avoid others wasting time. But I guess it's better if I don't lose my time plus I don't offend others.

Comment (MakeNowJust) [reply src/slice/sort.cr:404]: @asterite said in this thread before:

> Please let's use the names as in Rust, it'll be easier to apply patches.

This was removed, but I agree **this** opinion really. Anyone, what do you think?

File: src/slice/sort.cr
Comment (MakeNowJust) [reply src/slice/sort.cr:559]: I think so too. However, these lines are also ported from Rust completely (See [here](https://github.com/rust-lang/rust/blob/507bff92fadf1f25a830da5065a5a87113345163/library/alloc/src/slice.rs#L1132-L1137)). I'm worried whether this should be fixed or not in terms of maintainability.

@asterite or anyone, which do you like?

Comment (asterite) [reply src/slice/sort.cr:559]: In my experience leaving the code as similar as possible to the original one is better. Later on if the Rust code changes you can do a diff and see what changes you need to apply here. So please leave it like this.

And thank you for all your contributions, you are awesome! And I hope you are having fun coding in Crystal.

**Comment by MakeNowJust** [comment]
@j8r suggestions are accepted in a450923. Thank you!

I think all works are done now. Please review this!

File: src/slice/sort.cr
Comment (j8r) [reply src/slice/sort.cr:404]: I didn't have this insight. I even tought it can be a private instance method, but that's right, better let it a pure function similar to the Rust one 👍

**Comment by yxhuvud** [comment]
> The option's default value is false for backward compatibility.

Personally I'd not consider going from undefined order to defined order a backwards compatibility breaking change. Though perhaps I'm wrong there. However, I'd argue for that the default would be whatever is fastest. Is there any performance numbers? (performance tests can perhaps be found in the previous change to the sorting).

**Comment by asterite** [comment]
For Crystal, like in Ruby, I prefer correct over fast. In fact, if this stable sort is fast enough I would prefer to make stable the default and remove the non-stable option. There's less code and less worries for developers.

**Comment by MakeNowJust** [comment]
A rough benchmark:

```crystal
require "benchmark"

N = 100_000

puts "=== random ==="
rand = Random.new(seed: 42)
arr = [] of Int32
N.times { arr << rand.next_int }

Benchmark.ips do |x|
  x.report("stable: true") { arr.sort(stable: true) }
  x.report("stable: false") { arr.sort(stable: false) }
end

puts "=== partially sorted ==="
arr = [] of Int32
4.times do
  (N//4).times { |i| arr << i }
end

Benchmark.ips do |x|
  x.report("stable: true") { arr.sort(stable: true) }
  x.report("stable: false") { arr.sort(stable: false) }
end
```

And a result on my MacBook Air 2018 is:

```
=== random ===
 stable: true  93.78  ( 10.66ms) (± 6.09%)  587kB/op   1.57× slower
stable: false 147.20  (  6.79ms) (± 6.21%)  391kB/op        fastest
=== partially sorted ===
 stable: true   1.51k (662.69µs) (± 7.85%)  586kB/op        fastest
stable: false 306.63  (  3.26ms) (± 4.24%)  391kB/op   4.92× slower
```

I don't have a strong opinion, but I think we can switch to `stable: true` as the default.

**Comment by asterite** [comment]
Wow, the stable algorithm is pretty good! I think we should just leave that as a default.

**Comment by asterite** [comment]
Interesting, `sort` in Ruby is not stable: https://bugs.ruby-lang.org/issues/1089

But looking at the times above, it seems stable sort isn't that much slower... hmmm...

**Comment by Fryguy** [comment]
> The option's default value is false for backward compatibility.

I'd be surprised if anyone relied on instability of sorting, so is there even a backward compatibility concern?  As others have stated, IMO, default should be `stable: true`, with `stable: false` still available for possibly faster alternatives (if there even is a faster alternative).

**Comment by MakeNowJust** [comment]
Now, `stable: true` is the default.

File: spec/std/array_spec.cr
Comment (Sija) [inline spec/std/array_spec.cr:1343]: Use tuples over arrays for less amount of heap allocations:

```suggestion
    {true, false}.each do |stable|
```

ditto elsewhere

Comment (asterite) [reply spec/std/array_spec.cr:1343]: Doesn't matter at all in specs

**Comment by straight-shoota** [comment]
About specs: I would like to avoid quadruple tests for `Array#sort`, `Array#sort!`, `Slice#sort`, `Slice#sort!`. All delegate to the same implementation and there's no benefit for testing that multiple times. It's just a lot of unnecessary duplication.
I would prefer to have extensive specs only once, for the main implementation `Slice#sort!`. The other methods can be covered by simple ones testing only basic functionality.

**Comment by MakeNowJust** [comment]
@straight-shoota Please open another PR. Your concern can separate from this at all. This PR already contains two fixes (stable sort implementation and switching the default).

**Comment by straight-shoota** [comment]
It would be great to include the deduplication in this PR because it makes the spec changes much easier to review. But I can make a separate PR for that and then we merge it into this.

**Comment by Sija** [comment]
@straight-shoota Or you could include those within this PR, whatcha think @MakeNowJust?

**Comment by asterite** [comment]
I still think we should make `stable` be the only option for `sort`. That would mean there won't be duplicate specs.

**Comment by straight-shoota** [comment]
> That would mean there won't be duplicate specs.

I'm not talking about duplication stable vs. unstable behaviour but duplication between four different methods which all are based on the same implementation.

**Comment by MakeNowJust** [comment]
Removing existing specs is an incomprehensible thing in general. A squashed commit is hard to describe the reason. IMHO we should open another PR in this case.

**Comment by rdp** [comment]
Is anything holding this up, I'd be happy to help.

**Comment by straight-shoota** [comment]
Waiting on #10208.

**Comment by rdp** [comment]
What lacks still, does it need a rebase or something?  The failing check seems unrelated...thanks :)

**Comment by MakeNowJust** [comment]
Sorry, I forgot to rebase this. It's done now.

File: spec/std/array_spec.cr
Comment (straight-shoota) [inline spec/std/array_spec.cr:1346]: Maybe use an example list where `sort_by(&.size)` is not identical to `sort`?
Ditto for `sort_by!` and `Slice`.

File: spec/std/slice_spec.cr
Comment (straight-shoota) [inline spec/std/slice_spec.cr:629]: Please add the expected error message for clarity. Ditto below.

**Review by straight-shoota — APPROVED** [review]

**Comment by MakeNowJust** [comment]
Above two points by @straight-shoota are derived by the original spec. In short, they are not related to this actually, so we can fix them in another PR. (In fact, @straight-shoota approved this.)

@asterite @bcardiff  If anyone can review or merge this, please go it. It is ready.

**Comment by straight-shoota** [comment]
Should we merge this for 1.0? The default argument `stable = true` is technically a breaking change from the current unstable sort. I don't suppose it would have a practical impact if an unstable sort suddenly becomes stable (except for performance). So maybe it can wait for 1.1.

**Comment by MakeNowJust** [comment]
@straight-shoota Your comment says about "we can wait to merge", but, in a fair way, you should explain why we wait to merge. I'm not clear about this.

**Comment by straight-shoota** [comment]
We *could* wait to keep the delta between 0.36.1 and 1.0.0 small and delay potential surprises with unexpected side effects, for example. I'd like to merge it directly, though.

**Comment by bcardiff** [comment]
I would rather wait for 1.1 to avoid surprises like edge-cases, memory footprint. Or at least more heave benchmarks since this is changing the default sort.

As mentioned in https://github.com/crystal-lang/crystal/pull/10163#issuecomment-753962805 this can be seen as a non-breaking change so it can arrive after 1.0.

**Comment by straight-shoota** [comment]
While not a huge breaking chage, it is surely going to affect code that relies on the unstable sort implementation. Unstable is still deterministic.

Not sure if that's much relevant besides specs that test for specific sort results. But even that should be considered a breaking change.

**Comment by oprypin** [comment]
Why not merge it as off by default and later change to on by default

**Comment by oprypin** [comment]
^That way people get to try it out and tell about it, so maybe you won't need to guess whether it's breaks some expectation

**Comment by oprypin** [comment]
Actually that's not even my point.

It's a very useful feature to have, and the argument against merging it is only with the assumption that becomes the default.

**Comment by straight-shoota** [comment]
Why should it not be the default? I'd assume there is a mutual agreement that stable sort should be default. And if you don't need it but want more performance, you can disable it.

Adding (non-default) stable sort isn't relevant for 1.0. If the default behaviour doesn't change, it can just be introduced in 1.1. But changing default behaviour should better be in 1.0 or wait for 2.0.

**Comment by HertzDevil** [comment]
Heads up: this can now be applied on top of #10609.

**Comment by stakach** [comment]
might be worth merging this sooner rather than later given #10748

**Comment by Sija** [comment]
This could be merged already and the remaining optimization using #10609 could be applied in the followup PR - if @MakeNowJust has no time to do it right now.

**Comment by MakeNowJust** [comment]
#10609 is applied. This is ready to merge now.

File: src/compiler/crystal/tools/doc/type.cr
Comment (straight-shoota) [inline src/compiler/crystal/tools/doc/type.cr:169]: Since stable is default, the argument is technically not necessary. I don't mind keeping it explicit, though.

**Review by straight-shoota — APPROVED** [review]
This has already been milestoned for a long time, and I think it's ready for merging.

@MakeNowJust I would appreciate if you could just merge master instead of rebase. Force pushes make proper reviewing unnecessarily hard.

**Comment by asterite** [comment]
@straight-shoota Great!

Just to be sure: did you review the entire code and do you understand how it works? I mean, all the internals and details. I only ask because once this is merged, if there's a bug or something that needs to be changed, we must be able to fix it on our own. I didn't have time to review this yet. Or... if we aren't confident, we can mark this as experimental so we could remove it in the future if there are issues and we can't solve them.

File: src/compiler/crystal/tools/doc/type.cr
Comment (MakeNowJust) [reply src/compiler/crystal/tools/doc/type.cr:169]: Here's `stable` is meaningful and essential. When sorting implementation change other, it will be broken if `stable: true` is missing. Thus, I set `stable: true`.

@HertzDevil What's your opinion?

**Comment by straight-shoota** [comment]
In my recent review I just checked the diff between the previous HEAD and the force-pushed HEAD (which is really just the last commit 8647b6d), because I had already previously approved.

I don't think I completely understood every little detail. But the code looks readable, is well documented and tested. So I don't think there are should be any maintenance issues.

**Comment by straight-shoota** [comment]
In the last core team meeting we decided to postpone this PR after the release of 1.1 to rethink the public sorting API. See https://github.com/crystal-lang/crystal/issues/6057#issuecomment-861292171 for the discussion about that.
Sorry for the delay and thanks for your patience!

**Comment by straight-shoota** [comment]
Ooops, I think I was a bit too quick on this one. We were not satisfied with the API design.
Should've removed the milestone before 🙈 

But this is fine, probably. It brings the implementation and we can iterate on the API in a follow-up.

**Pull request merged**
